### DISCUSSION TITLE
Navigation mode drains battery faster than I can charge with my 2A charger

### DISCUSSION THREAD

**Discussion by black-square** [osmandapp/OsmAnd#11962] [General]
I have OsmAnd+ 2.7.2 on my Huawei Mate 9 phone and I cannot use it for a long drive navigation as it drains the battery faster than a charger can charge. Also, I noticed that the phone starts to heat up when I use OsmAnd with the charger connected. 

It's a good car 2A charger and it can really output 2A. I tested it by using AccuBattery app.

Also, I noticed that OsmAnd works less smoothly when the charger is disconnected which made me thinking that OsmAnd uses some sort of a battery economy technique and reduces redraw frequency. 

Is it possible to have an option (at least in the development plugin) to enable battery eco mode even when the charger is connected?

**Comment by sonora** [comment]
(1) Are you sure your device charges at all when the screen is on? Some device/charger combination do not do this.

(2) When your device is hot, it will probably no be charged in any case, this has nothing to do with OsmAnd. All Li batteries have a safety mechanism to stop charging from a certain temp upwards.

(3) Are you quite positive the issue is 100% correlated to the navigation mode being on? How do things differ if you just keep OsmAnd displaying the map without navigation?

OsmAnd does not use any special battery, eco etc. functionality, but instead relies on any OS feature to do this. If you e.g. tell your device to always throttle CPU usage ("Power savings mode"), or do this when on battery only, hat would explain your observations with regard to this.

**Comment by black-square** [comment]
Hi sonora,
(1) Yes, I'm pretty sure. I have AccuBattery app installed which allows displaying an overlay meter of battery current. When it's negative it means the discharge and positive value means charge. So, when I minimize OsmAnd or just stop the car  (navigation still active but doesn't render new tiles) the meter shows positive values. 

(2) Right, but I see that the discharge current exceed charge current before the device gets hot. Basically, it starts happening when I start moving. Although, during the night when I can set the brightness on the lowest level I can drive for hours and the battery is charging. The meter shows values less than 100 mA, but still positive. I cannot drive on min brightness during the day as the map becomes invisible.

(3) Sorry that I didn't clarify it before. I don't think that the problem is navigation mode by itself. It happens when the car is moving and the current position is pinned on screen. When the map has to render new tiles all the time that when it consumes so much power. The device is quite powerful and everything is perfectly smooth, but it consumes lots of energy, more than the charger can provide.

All battery saving features are unfortunately disabled in charging mode and I cannot limit the frequency of CPU which I would love to do but I cannot.

I agree that it's mostly my device problem. I think that the manufacturer didn't test much the device battery consumption on load in charging mode. Which is probably a quite rare case. The device shuts down all the power saving features and believes that power is always enough.

I think it would be useful if OsmAnd implements some of the power saving features. It could be very useful in battery mode as well. Modern devices are very powerful and many people will willingly sacrifice some visual smoothness if it leads to longer battery life. 

First of all, I would give an option to limit max screen update rate all the way up to 10 maybe even less FPS. Also, I didn't read the code, but I expect that you have a separate thread which renders raster tiles from a vector data. It would be nice to give an option to limit the minimum time between the start of the previous tile render job and the start of the new job. It should help a device to keep the frequency of CPU (and maybe even GPU) on low levels.

What do you think?

**Comment by sonora** [comment]
I am wondering if this is maybe related to #3898. There I reported huge battery and CPU consumption in connection with permanent address lookup. Are you perhaps only observing this for app modes where you have 'Display street name' selected in the Configure screen menu?

Can you look at some sort of Task manager and check CPU usage while this occurs? Somehow you should try to find out if it is the screen or the CPU which draws so much power.

Also temporarily disable all other and in particular 'permanent on' overlay etc. apps, they may themselves actually be the reason for causing this.

**Comment by black-square** [comment]
@sonora I've captured a video which I think quite interesting. My charger was always connected. AccuBattery monitored the battery current (bottom left corner) and the CPU load per each core (bottom right corner):

https://youtu.be/0Xp5F-S57gA

Time | Description
------------ | -------------
0:00 - 0:30 | I plugged the charger in and opened OsmAnd. The battery was charging with the current > 1.5A and CPU was almost idle. Also, the map was being updated in approximately once in a second.
0:30|I started the navigation mode.
0:31 - 1:09|Driving with navigation mode. Battery power consumption skyrocketed and the battery was discharging. The load of one of the CPU was 100%. Also, notice that the map refresh rate became very high. It looks very smooth and definitely > 20 FPS.
1:10|Disabled the navigation mode
1:11-1:24|Power consumption, CPU load, and screen refresh rate remained the same
1:24|Closed OsmAnd, killed the app and opened it again
1:28-1:57|Power consumption, CPU load, and screen refresh rate **still stayed the same**.
1:58|Closed OsmAnd and left the home screen opened
1:59-2:19|Battery started charging again, CPU is idle
2:20|Opened GoogleMaps
2:21-2:52|Google maps was moderately loading 4 cores. Battery was charging
2:53|Closed Google Maps and Opened OsmAnd
2:57-3:23|Battery was charging, CPU load is minor, refresh rate = 1 FPS
3:25|Entered the navigation mode
3:31-4:19|Battary started discharging, CPU core is red, rendering is perfectly smooth (FPS is very good)
4:20|Exited the navigation mode. Killed the OsmAnd process
4:24|Opened the Google Maps
4:38-5:00|Battery was charging

It feels like when I activate the navigation mode the device bumps up the CPU clock and the power consumption becomes very high. However, OsmAnd manages to utilize the full new CPU power and starts to render as often as possible and device keeps the CPU clock. Closing OsmAnd and giving some time to the device to idle or do moderate job brings the CPU back to the previous state until the navigation mode of OsmAnd is activated again.

So, I think that even if the navigation mode has some performance problems (which initiates the problem) OsmAnd should have an option to limit the maximum power consumption and reduce refresh rate to a reasonable level even if a device allows to render more frequently.

What do you think about that?

**Comment by scaidermern** [comment]
Does the battery consumption improve when you disable Configure screen -> Street name? As far as I understand this should disable the broken reverse geocoding from #3898. However I'm not an OsmAnd developer so I might be wrong.

**Comment by black-square** [comment]
@scaidermern I haven't tested it yet, but I will the next time I have a chance. But anyway, "Street name" could be a trigger, but the problem I think is deeper. Even after restart (1:24 on my video) OsmAnd keeps the device in a high-performance mode when the street name and navigation mode are disabled.

**Comment by sonora** [comment]
Chances are what you show around 1:24 is not a true "kill" of the OsmAnd app, seems to me the restart is way too fast for that. Tapping the (x) in the "recent apps" list on many Android versions does not really kill the app, but only sleeps it and/or puts things in the background. For a true "kill" to be achieved you will have to use some true task manager, or go to the Android's app settings and use "Force stop". 

Chances are the way you did it, reverse address lookups or other processes were actually continuing to run the background, causing continued battery drain (at least when the app was put in the foreground again).

**Comment by black-square** [comment]
@sonora I checked it with `adb shell ps | grep osmand` and can confirm that on my device tapping the "trash can" button or swiping the app thumbnail or tapping X in recent app menu does kill the OsmAnd process. Also, Mate 9 is pretty fast device and OsmAnd always starts pretty quick and you can notice as well that OsmAnd showed the splash screen which it doesn't do when it is activated from the background.

**Comment by black-square** [comment]
@scaidermern I tested it without street names and the result is pretty much the same. It might be slightly better, but not significantly.

**Comment by jbaber** [comment]
@black-square already described my situation exactly. I had to switch to google maps halfway through my trip so I'd have enough battery to get back home.

**Comment by sonora** [comment]
This is an old thread, and I could never reproduce it on my devices. Which device/Android/OsmAnd-version combinations does it still apply to?

**Comment by sonora** [comment]
One more experiment: Please enable the OsmAnd development plugin, then in its settings disable "Animate my position". Does this make a significant difference?

**Comment by jbaber** [comment]
I've got a Samsung Galaxy S9, SMG960U, Android 9 OSMAnd 3.3.8. I'll try that setting next time I navigate.

**Comment by kevenwyld** [comment]
Samsung S5 (klte) running LineageOS 16.0 and having the same issue.  @sonora's suggestion for turning off "Animate my position" made a big difference for me.  It was draining the battery at nearly 1% per minute before and now it charges, albeit quite slowly.  The map and navigation is still totally usable as well, the setting only disables the "smooth" movement animation of your position.

**Comment by oxxm** [comment]
Observing the same draining rate as in previous comment. Lately using Osmand navigation for longer cycling routes, and the battery is draining with a rate of nearly 1% per minute.

**Comment by benoiton** [comment]
I've got a Samsung Galaxy S7 (herolte), LineageOS 14.1 (Android 7.1.2), with OsmAnd~ 3.4.2. "Animate my position" is off. "No animation" is on. I use OS Monitor to see the CPU usage.
When moving (in a car), not centered (map fixed, blue circle moving), CPU consumption is about 3%. When centered (map moving, blue circle centered), CPU consumption is more than 30% (it's grows as I zoom out).
As a consequence, it's really drains my battery.

**Comment by sonora** [comment]
@benoiton That's great information thank you! A similar issue bothering me is #3898, I think we should verify if it is related or not. Could you run your test for the scenarios navigation case vs. non-navigation caae, and also street name widget visible vs. not?

PS: Animation settings seem now under General settings in latest release.

**Comment by benoiton** [comment]
Animation is always off.
I did not use navigation.
When not moving, CPU consumption is low (3%).
What is the street name widget?
(Map zoom is 150%, Text size is 125%)

**Comment by sonora** [comment]
The street name widget looks up the address you are at, displayed at the top of the map screen.

On by default in e.g. the car profile. It can e.g. in car profile be displayed/hidden via map screen/menu button/Configure screen/Street name. My suspicion is it may drain battery by constantly looking up the current address (if moving), hence I asked if it is on.

**Comment by benoiton** [comment]
I will check, but I'm pretty sure I was not in car profile.
(I will test again as soon as I am in a means of transportation with both hands free)

**Comment by benoiton** [comment]
I checked : CPU consumption is high, whether in car profile or not.
It seems to me that the more the screen moves, the higher the CPU consumption is. The CPU consumption is lower when the movement is strait than when turning (map rotation).

**Comment by sonora** [comment]
Looks like that indicates it is not the street name lookup and widget then.

If it is panning the map, I would expect turning off "Menu/Settings/General settings/Animate my Position" to make a noticeable difference?

**Comment by benoiton** [comment]
"Animate my Position" is already off.

**Comment by augustofagioli** [comment]
same issue on my Samsung S6. on same device,  
  I noticed also longer direction on   this last release  ( 200 km or more)  takes more than 5 minutes. none of these on prev release, same S6
Subscribing

**Comment by sonora** [comment]
Just to rule out something I have recently come across: Can I please ask people experiencing this to change the cable they use for charging? I have reently seen 2 Samsung devices exhibiting exactly this issue, and in both cases it went away miraculously after reverting to the original Samsung USB cable. Not sure yet what the deal is here, maybe the wire cross section is simply too narrow in some cables to deliver a decent charge....

**Comment by pinpox** [comment]
Same issue on Samsung S5 with lineageOS. Have tried various, good cables. Changing from a Anker powerbank that has more than enough power. I'm on a motorcyle and have read about vibrations reducing the charging rate. Just ordered a magnetic cable that is supposed to help with that issue. If it helps, I'll report back but I don't expect miracles from it. Also to rule out a bad battery I have tried 3 different ones too (S5 has removable battery)

**Comment by benoiton** [comment]
A good cable is important for fast charging. It can help car drivers. But there shouldn't be such power drain at the first place...
I don't have a power supply on my bike. My brother can use OsmAnd on his Samsung Galaxy A8 for several hours when biking. I cannot. It's weird!

**Comment by Ady10101** [comment]
I am also having the same problem with omsand using more power than the charger can keep up with on my motorcycle. Not a problem with Google maps though.

**Comment by Rikests** [comment]
I have been dealing with this problem too. I get very high power consumption and phone heating when screen is on and the map is following my movements. The animations are smooth, but my car charger cant keep up with the power consumption and my phone slowly discharges, even while connected  to the charger.
In previous versions (around a year ago) i could avoid this by ''bugging'' the osmand map in to a low fps mode. I did that by quickly scrolling the map in random directions and recentering it before it loads the new roads. Took a few tries, but eventually it limited the map updates to around 1 or 2 fps. I can not do this trick in current versions.
For now my solution to this is, navigating using notifications on the lockscreen.
I think this could be solved by adding a map update rate limiter, something like the fps limmiters present in modern games.

https://files.fm/u/kekbcu7q video will be deleted after 2 months.
In the video, you can see how the power consumption goes down from ~1000mA to ~300mA when the map is static and not centered on the current location. FYI I was over the sea in an airplaine while making this video, but exactly the same problem happens when im on the ground, driving or walking.

Im using oneplus X with LineageOS

**Comment by benoiton** [comment]
I updated my Galaxy S7 from LineageOS 14.1 (Android 7.1) to 16 (Android 9, unofficial for this model). And there is no more any battery drain. For instance, I used yesterday for a 40' bike ride, with screen on and navigation, and the battery lost just 10%. Before that, it would have been maybe 60%.
OsmAnd~ 3.5.5.
Therefore, at least for me, it's neither a hardware issue nor bound to some hardware.
I supposed some graphic lib does not use hardware drawing.
LineageOS seems highly represented in this thread.

**Comment by urld** [comment]
I experience the same issue on a Google Pixel, Android 10, OsmAnd 3.5.5.
The phone gets really hot and the battery gets drained quickly even while charging. When using Google Maps, the battery usually reaches it's full charge until I arrive at the destination.

**Comment by VIMVa** [comment]
Same here on Nokia 9 pureview (Android 10).
My powerbank (when biking/hiking) nor car charger can follow the battery drain and phone gets extremely hot when navigating with OsmAnd+.
Tried all above suggestions.
Only when navigating, not when simply looking at the map.
Locus Maps (also with OSM maps), Google maps, Waze, TomTom GO,... do not show that behaviour at all.
OsmAnd is of no use for navigation.

**Comment by TobiWaterfield** [comment]
Just a brief "me too" with a test:

OsmAnd displaying the map with GPS and mobile data enabled:
After 4 hours 35% battery remain.

OsmAnd in **navigation** mode with GPS and mobile data enabled:
After 40 minutes 55% battery remain.

OsmAnd 3.6.3
Hardware: Samsung Galaxy S5 (SM-G900F)
Android version: 9
OS: LineageOS 16

Edit - Perhaps unrelated, but just to add to the potential knowledge base:
After the phone was quickly drained from 100% to 55% running OsmAnd, it actually stayed at 55% for the next 4½ hours while NOT running OsmAnd. The phone only saw normal/light usage in those 4½ hours, but certainly the battery level should have dropped. I have never seen this behavior before. Perhaps something is going on with how LineageOS/Android reports the battery level that is triggered by OsmAnd.

**Comment by lydenyardley** [comment]
Same for me. Phone gets ridiculously hot and drains at least 1% per min in navigation.

Galaxy S9 G960F
Android 10
LineageOS 17.1 no gapps or even microG
Latest OsmAnd

Basically unusable for long journeys :(

Any progress?

**Comment by lefou** [comment]
Same for me. just recording a track needs almost no power, but once navigation is active, battery drains really fast and phone gets hot.

**Comment by lefou** [comment]
This is reproducible in navigation simulation mode.

**Comment by DamianoPantani** [comment]
Same here, this is perhaps the most critical issue with the app, and I'm concerned it's not being looked at for this many years now.

**EDIT: FOUND THE PROBLEM, SOLUTION BELOW**

Disabling street names and "No animations -> on" didn't change a thing for me, battery was still discharging with 10W wireless charger on, at about 10% an hour.

But in the middle of a trip I went to **Settings -> (Profile of your choice) -> Navigation settings -> Animate own position -> OFF**. Battery started charging at about 20% an hour. Massive difference.

**Comment by sonora** [comment]
@vshcherb Intetesting! If this affects more devices, I guess we should make this our installation default.

**Comment by nils-werner** [comment]
It makes sense that this setting is more power hungry, as it will cause hundreds more map redraws to do one second of animation.

I think two things could be investigated here:

1. Profile the renderer code and find segments that are especially power hungry
2. Make the bitmap cache smarter so that "own position" can be animated on top of the bitmap data, and not trigger a complete redraw hundreds of times per second.

@sonora I think disabling animations is just a quick workaround but not a fix for this issue.

**Comment by vshcherb** [comment]
It's expected but it still didn't give enough power in the end. If you think that Animation causes redrawing 5-10 FPS and without this option map is updated only 1 FPS, then 20% improve doesn't look very impressive and definitely has big drawback how navigation feels.

**Reply by DamianoPantani** [reply]
It went from 10% **discharging** to 20% **charging**, so IMO it's a 300% improvement

**Reply by vshcherb** [reply]
Good point, I misread it.

**Reply by sonora** [reply]
@vshcherb It should behave roughly linear with FPS, I guess, meaning 1 FPS instead of 5-10 FPS could be in the ballpark of 80 - 90% saving(!)

**Comment by nils-werner** [comment]
I just came across [Streetcomplete](https://github.com/streetcomplete/StreetComplete/) and it's map is much more fluid and responsive, and uses less battery.

Maybe it's rendering code can serve as an inspiration.
